### PR TITLE
Add testing for remove_segment

### PR DESCRIPTION
**Pull request by joaopapereira**
This PR will add some testing around remove_segment and make the code more explicit. We found that the `remove_segment` was calling `get_seginfo` and ignoring the the output because it was piggy backing on that function to call `elog` if the segment did not exist.

To solve this we created a new function called `does_segment_exist` that will check if the segment exit.

Also in this PR improves cmockery functions to make the errors more explicit and add a function that allow a test to fail and write a custom message

### CODE REVIEW

File: src/backend/cdb/cdbutil.c [hunk 1]
```diff
@@ -1034,6 +1034,42 @@ master_standby_dbid(void)
 	return dbid;
 }
 
+bool
+does_segment_exist(int16 dbid)
```
Comment (danielgustafsson) [inline src/backend/cdb/cdbutil.c]: This is duplicating code (to the point where the comment in the `elog()` below isn't updated) from `dbid_get_dbinfo()`. That function is in turn used in a single place, and that single place is called in only two places. Can't we instead refactor `get_seginfo()`to `dbid_get_dbinfo(int16 dbid, bool errorOK)` and have it return `NULL` rather than call `elog(ERROR..` on failures. `get_seginfo()` can then pass in `false`, and this can either become a thin wrapper passing in `true` or simply go away with this usecase calling `dbid_get_dbinfo(.., true)` directly. Yes, it will allocate a little memory for no use, but that's cheap compared to carrying duplicate code only for a test.

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -3894,7 +3894,6 @@ groupMemOnDumpForCgroup(ResGroupData *group, StringInfo str)
 
 /*
  * Parse cpuset to bitset
- * if onlyCheck is true, the function only check whether cpuset is valid
```
Comment (danielgustafsson) [inline src/backend/utils/resgroup/resgroup.c]: This seems unrelated to the PR? Please push this separately, and rebase out of the PR. There is no need for a PR for this commit as it's clearly an incorrect comment.

File: src/test/unit/cmockery/cmockery.c [hunk 3]
```diff
@@ -1402,7 +1402,16 @@ void _assert_true(const LargestIntegralType result,
...
+		print_error(OUTPUT_PADDING "expected '%s' to be true\n", expression);
+		_fail(file, line);
+	}
+}
+
+void _assert_false(const LargestIntegralType result,
+		const char * const expression,
+		const char * const file, const int line) {
+	if (!result) {
+		print_error(OUTPUT_PADDING "expected '%s' to be false\n", expression);
```
Comment (danielgustafsson) [inline src/test/unit/cmockery/cmockery.c]: The commit message for these should mention that they've been submitted upstream and contain the upstream commit sha from the cmockery repo. In the unlikely event that they ever put out another release, we need to know the relation to local commits.

File: src/backend/cdb/test/cdbutil_test.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+ * back to the next available PG_CATCH();
+ */
+void
+_ExceptionalCondition()
+{
+	PG_RE_THROW();
+}
+
+static void
+expect_elog(int log_level)
```
Comment (danielgustafsson) [inline src/backend/cdb/test/cdbutil_test.c]: Why is this duplicated in the `elog_helper`?

File: src/backend/access/appendonly/test/aomd_test.c [hunk 5]
```diff
@@ -182,9 +181,8 @@ test_mdunlink_co_4_columns_1_concurrency(void **state)
 
 	mdunlink_ao(PATH_TO_DATA_FILE);
 
-	assert_true(num_unlink_called == 4);
+	assert_int_equal(num_unlink_called, 4);
```
Comment (danielgustafsson) [inline src/backend/access/appendonly/test/aomd_test.c]: Why are we changing from assert_true to assert_int_equal? (As in, can you elaborate on why this is more descriptive than asserting that a statement is true?)

File: src/backend/cdb/test/cdbutil_test.c [hunk 6]
```diff
@@ -0,0 +1,4 @@
+//
+// Created by pivotal on 9/24/18.
+//
+
```
Comment (danielgustafsson) [inline src/backend/cdb/test/cdbutil_test.c]: Added by mistake?

File: src/backend/cdb/cdbutil.c (hunk 1 above)
Comment (danielgustafsson) [reply src/backend/cdb/cdbutil.c]: (reading a bit more) Is this adding a function whose purpose is to make mocking it easier?

File: src/test/unit/cmockery/cmockery.c (hunk 3 above)
Comment (joaopapereira) [reply src/test/unit/cmockery/cmockery.c]: Done

File: src/backend/access/appendonly/test/aomd_test.c (hunk 5 above)
Comment (joaopapereira) [reply src/backend/access/appendonly/test/aomd_test.c]: Updated the commit message. The the gist of it is if you use assert_true the error output will be:
```
 expected 'num_unlink_called == 4' to be false
```

while if you use assert_int_equal:
```
1 != 4
```
eventually we can also improve this output and make it more readable

File: src/backend/cdb/test/cdbutil_test.c (hunk 6 above)
Comment (joaopapereira) [reply src/backend/cdb/test/cdbutil_test.c]: yes, i removed it

**Comment by danielgustafsson** [comment]
> On 27 Sep 2018, at 17:24, João Pereira <notifications@github.com> wrote:
> 
> @joaopapereira commented on this pull request.
> 
> In src/backend/utils/resgroup/resgroup.c <https://github.com/greenplum-db/gpdb/pull/5872#discussion_r220969076>:
> 
> > @@ -3894,7 +3894,6 @@ groupMemOnDumpForCgroup(ResGroupData *group, StringInfo str)
>  
>  /*
>   * Parse cpuset to bitset
> - * if onlyCheck is true, the function only check whether cpuset is valid
> You are right, but a PR for this removal looks a bit overkill. Nevertheless this was isolated in one commit
> 
It would’ve been overkill yes, but you don’t need to open a PR for trivial comment fixes, just use your own judgment and push them directly.

**Comment by joaopapereira** [comment]
> Looking at the remove_segment() function, the call to get_seginfo() is useless. The comment "Check that we're removing a mirror, not a primary" is not even valid since get_seginfo() -> dbid_get_dbinfo() call stack doesn't even check for segment role.

That is exactly what we are replacing in this PR. We remove the `dbid_get_dbinfo` call and start using a new function that just checks if the segment exists. In this function we are doing the only check that `dbid_get_dbinfo` did, see if the dbid exist and if not `elog`

> Or better yet, delete remove_segment(), make remove_segment_config() do a return numDel > 0, and have gp_remove_segment(), gp_remove_segment_mirror(), and gp_remove_master_standby() functions directly call remove_segment_config() instead. With gp_remove_segment() now returning a boolean, we can actually send true/false back to the front-end user/utility instead of always returning true or erroring out.

So your suggestion here is to change the behavior that we currently have (erroring out when cannot find the mirror id) and start returning a boolean?

File: src/backend/access/appendonly/test/aomd_test.c (hunk 5 above)
Comment (danielgustafsson) [reply src/backend/access/appendonly/test/aomd_test.c]: Since the integer output doesn't contain the variable name I would find the other output easier to read since `1  != 4` doesn't expose the extra information of the variable name that `num_unlink_called == 4` does.

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (danielgustafsson) [reply src/backend/utils/resgroup/resgroup.c]: Yeah, as stated elsewhere, no need for a PR for comment fixups, just push directly.

**Comment by asimrp** [comment]
I agree with Jimmy's suggestion.  I think this is what his "simpler" suggestion would look like:
```diff
diff --git a/src/backend/utils/gp/segadmin.c b/src/backend/utils/gp/segadmin.c
index 15b39eaeba..3d07c92b25 100644
--- a/src/backend/utils/gp/segadmin.c
+++ b/src/backend/utils/gp/segadmin.c
@@ -298,39 +298,6 @@ add_segment_config_entry(int16 pridbid, seginfo *i)
 	add_segment_config(i);
 }
 
-/*
- * Remove a gp_segment_configuration entry
- */
-static void
-remove_segment_config(int16 dbid)
-{
-	int			numDel = 0;
-	ScanKeyData scankey;
-	SysScanDesc sscan;
-	HeapTuple	tuple;
-	Relation	rel;
-
-	rel = heap_open(GpSegmentConfigRelationId, RowExclusiveLock);
-
-	ScanKeyInit(&scankey,
-				Anum_gp_segment_configuration_dbid,
-				BTEqualStrategyNumber, F_INT2EQ,
-				Int16GetDatum(dbid));
-
-	sscan = systable_beginscan(rel, GpSegmentConfigDbidIndexId, true,
-							   SnapshotNow, 1, &scankey);
-	while ((tuple = systable_getnext(sscan)) != NULL)
-	{
-		simple_heap_delete(rel, &tuple->t_self);
-		numDel++;
-	}
-	systable_endscan(sscan);
-
-	Assert(numDel > 0);
-
-	heap_close(rel, NoLock);
-}
-
 static void
 add_segment(seginfo new_segment_information)
 {
@@ -488,12 +455,31 @@ gp_add_segment(PG_FUNCTION_ARGS)
 static void
 remove_segment(int16 pridbid, int16 mirdbid)
 {
-	seginfo    *i;
+	int			numDel = 0;
+	ScanKeyData scankey;
+	SysScanDesc sscan;
+	HeapTuple	tuple;
+	Relation	rel;
 
-	/* Check that we're removing a mirror, not a primary */
-	i = get_seginfo(mirdbid);
+	rel = heap_open(GpSegmentConfigRelationId, RowExclusiveLock);
 
-	remove_segment_config(mirdbid);
+	ScanKeyInit(&scankey,
+				Anum_gp_segment_configuration_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(mirdbid));
+
+	sscan = systable_beginscan(rel, GpSegmentConfigDbidIndexId, true,
+							   SnapshotNow, 1, &scankey);
+	while ((tuple = systable_getnext(sscan)) != NULL)
+	{
+		simple_heap_delete(rel, &tuple->t_self);
+		numDel++;
+	}
+	systable_endscan(sscan);
+
+	elogif(numDel > 0, ERROR, "could not find configuration entry for dbid %i", mirdbid);
+
+	heap_close(rel, NoLock);
 }
 
 /*
```

I also see that a unit-test for such a function is overkill.  There is hardly any logic in `remove_segment` function besides invoking heap access method interface.  If we are to write such unit-tests, why leave out `dbid_get_dbinfo`?  As an example of functions worthy of unit-tests, I would suggest looking at `src/backend/fts`.

File: src/backend/access/appendonly/test/aomd_test.c (hunk 5 above)
Comment (joaopapereira) [reply src/backend/access/appendonly/test/aomd_test.c]: I understand that, nevertheless the only information you get from the first one is that `num_unlink_called` doesn't have the value 4, this is an ok information but it is far more relevant knowing the real value of `num_unlink_called`. Because after this we can go to the test and see what is being checked.

**Comment by joaopapereira** [comment]
> I also see that a unit-test for such a function is overkill. 

Perhaps it is overkill due to the size of the function, but now we know (if we read the test) that when you call the function `remove_segment` you might get an `elog` error instead of having to go through all the code paths in order to see it.
To be fair this is one of the biggest advantages of unit testing, by reading the test you are able to understand what to expect of a function, and not have to read the full code to understand what the function is doing.

---

Thanks for all the cool ideas on how to improve this PR.
My suggestion for this is to take an iterative approach. We merge this PR and create a Chore in our backlog to see how we can improve this piece of code even more.

Are y'll ok with the suggestion?

**Comment by asimrp** [comment]
The feedback so far falls more on the side of -1.  I would then suggest closing the PR and creating an issue as a followup.

**Comment by jimmyyih** [comment]
I would feel uneasy committing the main part of this PR in (the `does_segment_exist()` function and `remove_segment()` refactor).  If we take my suggestion or Asim's diff as the next iterative approach, it would essentially be reverting a large part of this PR's commit and applying the new idea.

**Comment by joaopapereira** [comment]
As I said before the proposed change should be a story in our backlog specially if it includes changing interfaces and other tools. In order to honor that I am going to create a Story with this information and tag this PR and close it without merging it.

All the changes that are not related with these proposed change I will gather them together and pushed them to Master as I do believe they are completely unrelated to the discussion that we are having above
